p2p/src/authenticated/actors/dialer.rs

```diff
@@ -147,15 +146,45 @@ impl<E: Spawner + Clock + GClock + Network + Rng + CryptoRng + Metrics, C: Schem
         mut tracker: tracker::Mailbox<E, C>,
         mut supervisor: spawner::Mailbox<E, SinkOf<E>, StreamOf<E>, C::PublicKey>,
     ) {
+        let mut dial_deadline = self.context.current();
+        let mut query_deadline = self.context.current();
         loop {
-            // Attempt to dial peers we know about
-            self.dial_peers(&mut tracker, &mut supervisor).await;
-
-            // Sleep for a random amount of time up to the dial frequency
-            let wait = self
-                .context
-                .gen_range(Duration::default()..self.dial_frequency);
-            self.context.sleep(wait).await;
+            select! {
+                _ = self.context.sleep_until(dial_deadline) => {
+                    // Update the deadline.
+                    dial_deadline = self.random_wait(self.dial_frequency);
+
+                    // Pop the queue until we can reserve a peer.
+                    // If a peer is reserved, attempt to dial it.
+                    while let Some(peer) = self.queue.pop() {
+                        // Attempt to reserve peer.
+                        let Some(reservation) = tracker.dial(peer).await else {
+                            continue;
+                        };
+                        self.dial_peer(reservation, &mut supervisor).await;
+                    }
+                },
+                _ = self.context.sleep_until(query_deadline) => {
```

Comment:
I wonder if it is optimal to just pop until empty (and only "hard refresh" in the case a new peer set is added)? I'm not loving this query refresh loop (but the other change is nice).

Reply:
I don't think we can do that since we need some way to try and dial peers that disconnect, even if there is no peer set added

Reply:
I suppose the approach you have here prevents us from calling dialable a bunch (when there is a small number of unconnected peers).

🤔

Reply:
Yes I wanted to not call `dialable` on repeat if there are a small number of unconnected peers.

I do see that we may be slow to connect to peers introduced in new peer sets if a peer set is added. However, also remember that we will not instantly have the addresses of new peers anyway since it takes some time to discover them. More likely than not, new peers will be the ones to dial existing peers since they already should have all the existing peer information